File: python/tests/test_elastic_bands.py
"""Elastic Bands (Quinlan & Khatib 1993): deforms a bubble chain draped over a
reference path away from obstacles each tick, via the closed-loop simulator on
a real map (obstacles the raw reference path cuts straight through) and via
direct compute_command calls for the deformation/maintenance behavior a full
episode can't isolate cleanly (straight-line equilibrium, band breaking)."""

from __future__ import annotations

import io
import json
from pathlib import Path
from typing import Any

import pytest
import yaml
from conftest import REPO_ROOT

from navigation.core.params import ParamError, ParamSet
from navigation.core.trace import TraceRecorder
from navigation.core.types import LocalTask, RobotState
from navigation.local_planning.band._band import resample_polyline
from navigation.local_planning.band.elastic_bands import ElasticBandsPlanner
from navigation.local_planning.simulation import SimConfig, SimStatus, simulate
from navigation.maps.loader import load_map, load_scenario

_ALGO = "elastic_bands"
_CONFIG_PATH = REPO_ROOT / "configs" / "local_planning" / f"{_ALGO}.yaml"
_MAPS_DIR = REPO_ROOT / "maps"


def _config(tmp_path: Path, **overrides: Any) -> ParamSet:
    """Real config with selected defaults overridden."""
    doc = yaml.safe_load(_CONFIG_PATH.read_text())
    for entry in doc["params"]:
        if entry["name"] in overrides:
            entry["default"] = overrides[entry["name"]]
    out = tmp_path / f"{_ALGO}.yaml"
    out.write_text(yaml.safe_dump(doc))
    return ParamSet.from_yaml(out)


def _sim_config(params: ParamSet) -> SimConfig:
    return SimConfig(
        control_dt=params.get_float("control_dt"),
        max_steps=params.get_int("max_steps"),
        goal_tolerance=params.get_float("goal_tolerance"),
        footprint_radius=params.get_float("footprint_radius"),
        stall_window=params.get_int("stall_window"),
        stall_distance=params.get_float("stall_distance"),
    )


def _run(map_name: str, scenario_name: str, params: ParamSet) -> Any:
    grid = load_map(_MAPS_DIR / "grid" / f"{map_name}.yaml")
    scenario = load_scenario(_MAPS_DIR / "scenarios" / f"{scenario_name}.yaml")
    planner = ElasticBandsPlanner(params)
    start = RobotState(pose=(scenario.start[0], scenario.start[1], scenario.start_theta))
    task = LocalTask(
        goal=(scenario.goal[0], scenario.goal[1], scenario.goal_theta),
        reference_path=scenario.reference_path,
    )
    return simulate(planner, grid, start, task, _sim_config(params), None)


def _band_events(buf: io.StringIO) -> list[dict[str, Any]]:
    events = [json.loads(line) for line in buf.getvalue().splitlines()]
    return [e for e in events if e["event"] == "band_updated"]


# --- (a) the raw reference path cuts through 5 obstacle blocks; the deformed --
# --- band routes around all of them without collision -------------------------
def test_reaches_goal_around_blocking_obstacles(tmp_path: Path) -> None:
    params = _config(tmp_path)
    result = _run("clutter01", "clutter01_s3", params)
    assert result.status is SimStatus.REACHED
    assert result.success is True
    assert result.min_clearance > 0.0


# --- (b) after one tick's deformation, the band's worst-case clearance beats --
# --- the raw (undeformed) reference path's worst-case clearance ---------------
def test_band_deforms_away_from_obstacle(tmp_path: Path) -> None:
    params = _config(tmp_path)
    grid = load_map(_MAPS_DIR / "grid" / "clutter01.yaml")
    scenario = load_scenario(_MAPS_DIR / "scenarios" / "clutter01_s3.yaml")
    planner = ElasticBandsPlanner(params)
    state = RobotState(pose=(scenario.start[0], scenario.start[1], scenario.start_theta))
    task = LocalTask(
        goal=(scenario.goal[0], scenario.goal[1], scenario.goal_theta),
        reference_path=scenario.reference_path,
    )

    spacing = params.get_float("bubble_spacing")
    baseline_points = resample_polyline(list(scenario.reference_path), spacing)
    baseline_min_clearance = min(grid.distance_to_nearest(p) for p in baseline_points)

    buf = io.StringIO()
    recorder = TraceRecorder(buf)
    planner.compute_command(grid, state, task, params.get_float("control_dt"), recorder)

    band = _band_events(buf)[-1]["band"]
    # radius = min(distance_to_nearest, rho_max), already computed per bubble.
    deformed_min_clearance = min(item[2] for item in band)

    assert deformed_min_clearance > baseline_min_clearance


# --- (c) an obstacle-free straight corridor: contraction alone keeps the band -
# --- exactly on the line joining its two fixed endpoints ----------------------
def test_straight_band_stays_straight(tmp_path: Path) -> None:
    params = _config(tmp_path)
    grid = load_map(_MAPS_DIR / "grid" / "open01.yaml")
    start_xy = (2.0, 8.75)
    goal_xy = (8.0, 8.75)  # horizontal segment verified obstacle-free within rho_influence
    planner = ElasticBandsPlanner(params)
    state = RobotState(pose=(start_xy[0], start_xy[1], 0.0))
    task = LocalTask(goal=(goal_xy[0], goal_xy[1], 0.0), reference_path=(start_xy, goal_xy))

    buf = io.StringIO()
    recorder = TraceRecorder(buf)
    planner.compute_command(grid, state, task, params.get_float("control_dt"), recorder)

    band = _band_events(buf)[-1]["band"]
    for _cx, cy, _rho in band:
        assert cy == pytest.approx(start_xy[1], abs=1e-6)


# --- (d) an unreachable rho_min forces the very first tick's validity check ---
# --- to fail: zero command, broken flagged in the trace ------------------------
def test_band_break_yields_zero_command(tmp_path: Path) -> None:
    params = _config(tmp_path, rho_min=1.2)
    grid = load_map(_MAPS_DIR / "grid" / "clutter01.yaml")
    scenario = load_scenario(_MAPS_DIR / "scenarios" / "clutter01_s3.yaml")
    planner = ElasticBandsPlanner(params)
    state = RobotState(pose=(scenario.start[0], scenario.start[1], scenario.start_theta))
    task = LocalTask(
        goal=(scenario.goal[0], scenario.goal[1], scenario.goal_theta),
        reference_path=scenario.reference_path,
    )

    buf = io.StringIO()
    recorder = TraceRecorder(buf)
    cmd = planner.compute_command(grid, state, task, params.get_float("control_dt"), recorder)

    assert cmd.v == 0.0
    assert cmd.omega == 0.0
    events = _band_events(buf)
    assert len(events) == 1
    assert events[0]["data"]["broken"] == 1.0


# --- (e) an out-of-range override fails load-time validation ------------------
def test_param_validation(tmp_path: Path) -> None:
    with pytest.raises(ParamError):
        _config(tmp_path, k_repulsion=-1.0)

File: cpp/tests/test_elastic_bands.cpp
#include <algorithm>
#include <cmath>
#include <fstream>
#include <limits>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/core/params.hpp"
#include "navigation/core/trace.hpp"
#include "navigation/core/types.hpp"
#include "navigation/local_planning/band/band.hpp"
#include "navigation/local_planning/band/elastic_bands.hpp"
#include "navigation/local_planning/simulation.hpp"
#include "navigation/maps/loader.hpp"
#include "navigation/maps/occupancy_grid.hpp"
#include "test_util.hpp"

// Elastic Bands (Quinlan & Khatib 1993): deforms a bubble chain draped over a
// reference path away from obstacles each tick, via the closed-loop simulator
// on a real map (obstacles the raw reference path cuts straight through) and
// via direct compute_command calls for the deformation/maintenance behavior a
// full episode can't isolate cleanly (straight-line equilibrium, band
// breaking).

using namespace navigation;
using core::LocalTask;
using core::ParamSet;
using core::Point;
using core::Pose;
using core::RobotState;
using core::TraceRecorder;
using local_planning::ElasticBandsPlanner;
using local_planning::resample_polyline;
using local_planning::SimConfig;
using local_planning::SimStatus;
using local_planning::simulate;

namespace {

std::string config_path() { return test::repo_path("configs/local_planning/elastic_bands.yaml"); }

ParamSet real_config() { return ParamSet::from_yaml(config_path()); }

ParamSet config_with_rho_min(double rho_min) {
  std::ostringstream extra;
  extra << "\n  - {name: rho_min, type: float, default: " << rho_min
        << ", min: 0.05, max: 2.0, description: override}\n";
  std::string base = test::repo_path("configs/local_planning/elastic_bands.yaml");
  std::ifstream in(base);
  std::ostringstream buf;
  buf << in.rdbuf();
  std::string doc = buf.str();
  // Overriding rho_min by re-declaring it: ParamSet::from_yaml keeps the last
  // occurrence of a repeated key when the yaml parser folds duplicate map
  // keys, mirroring _config()'s python-side "rewrite the entry" override.
  std::string path = test::write_temp("elastic_bands_rho_min.yaml", doc + extra.str());
  return ParamSet::from_yaml(path);
}

SimConfig sim_config_from(const ParamSet& params) {
  return SimConfig{params.get_float("control_dt"),     params.get_int("max_steps"),
                   params.get_float("goal_tolerance"), params.get_float("footprint_radius"),
                   params.get_int("stall_window"),     params.get_float("stall_distance")};
}

maps::OccupancyGrid2D& as_grid(core::MapBase& m) {
  return *dynamic_cast<maps::OccupancyGrid2D*>(&m);
}

std::vector<std::string> split_lines(const std::string& s) {
  std::vector<std::string> out;
  std::istringstream in(s);
  std::string line;
  while (std::getline(in, line)) out.push_back(line);
  return out;
}

std::vector<std::string> lines_with_event(const std::vector<std::string>& lines, const char* event) {
  std::vector<std::string> out;
  std::string needle = std::string("\"") + event + "\"";
  for (const std::string& line : lines) {
    if (line.find(needle) != std::string::npos) out.push_back(line);
  }
  return out;
}

double find_data_value(const std::string& line, const std::string& key) {
  std::string needle = "\"" + key + "\":";
  size_t at = line.find(needle);
  if (at == std::string::npos) return std::nan("");
  size_t start = at + needle.size();
  size_t end = line.find_first_of(",}", start);
  return std::stod(line.substr(start, end - start));
}

// Parses the `"band":[[x,y,r],...]` array out of one raw trace line. Manual
// bracket-depth scanning (no JSON library in this codebase's C++ tests --
// find_data_value above is the established precedent for reading trace JSON
// straight out of the raw line).
std::vector<std::vector<double>> parse_band(const std::string& line) {
  std::string needle = "\"band\":[";
  size_t start = line.find(needle);
  if (start == std::string::npos) return {};
  size_t outer_open = start + needle.size() - 1;
  int depth = 0;
  size_t outer_close = std::string::npos;
  for (size_t i = outer_open; i < line.size(); ++i) {
    if (line[i] == '[') ++depth;
    else if (line[i] == ']') {
      --depth;
      if (depth == 0) {
        outer_close = i;
        break;
      }
    }
  }
  std::vector<std::vector<double>> result;
  std::string inner = line.substr(outer_open + 1, outer_close - outer_open - 1);
  size_t j = 0;
  while (j < inner.size()) {
    if (inner[j] != '[') {
      ++j;
      continue;
    }
    size_t close = inner.find(']', j);
    std::string item = inner.substr(j + 1, close - j - 1);
    std::vector<double> vals;
    std::istringstream ss(item);
    std::string tok;
    while (std::getline(ss, tok, ',')) vals.push_back(std::stod(tok));
    result.push_back(vals);
    j = close + 1;
  }
  return result;
}

}  // namespace

// (a) the raw reference path cuts through 5 obstacle blocks; the deformed
// band routes around all of them without collision -----------------------------
TEST(ElasticBands, ReachesGoalAroundBlockingObstacles) {
  auto params = real_config();
  auto map = maps::load_map(test::repo_path("maps/grid/clutter01.yaml"));
  auto& grid = as_grid(*map);
  maps::Scenario sc = maps::load_scenario(test::repo_path("maps/scenarios/clutter01_s3.yaml"));
  ElasticBandsPlanner planner(params);
  RobotState start{Pose{sc.start.x, sc.start.y, sc.start_theta}, 0.0, 0.0};
  LocalTask task{Pose{sc.goal.x, sc.goal.y, sc.goal_theta}, sc.reference_path};
  auto result = simulate(planner, grid, start, task, sim_config_from(params), nullptr);
  EXPECT_EQ(result.status, SimStatus::REACHED);
  EXPECT_TRUE(result.success);
  EXPECT_GT(result.min_clearance, 0.0);
}

// (b) after one tick's deformation, the band's worst-case clearance beats the
// raw (undeformed) reference path's worst-case clearance -----------------------
TEST(ElasticBands, BandDeformsAwayFromObstacle) {
  auto params = real_config();
  auto map = maps::load_map(test::repo_path("maps/grid/clutter01.yaml"));
  auto& grid = as_grid(*map);
  maps::Scenario sc = maps::load_scenario(test::repo_path("maps/scenarios/clutter01_s3.yaml"));
  ElasticBandsPlanner planner(params);
  RobotState state{Pose{sc.start.x, sc.start.y, sc.start_theta}, 0.0, 0.0};
  LocalTask task{Pose{sc.goal.x, sc.goal.y, sc.goal_theta}, sc.reference_path};

  std::vector<Point> baseline_points = resample_polyline(sc.reference_path, params.get_float("bubble_spacing"));
  double baseline_min_clearance = std::numeric_limits<double>::infinity();
  for (const Point& p : baseline_points) {
    baseline_min_clearance = std::min(baseline_min_clearance, grid.distance_to_nearest(p));
  }

  std::ostringstream os;
  TraceRecorder rec(os);
  planner.compute_command(grid, state, task, params.get_float("control_dt"), &rec);

  std::vector<std::string> band_lines = lines_with_event(split_lines(os.str()), "band_updated");
  ASSERT_FALSE(band_lines.empty());
  std::vector<std::vector<double>> band = parse_band(band_lines.back());
  ASSERT_FALSE(band.empty());
  double deformed_min_clearance = std::numeric_limits<double>::infinity();
  for (const auto& item : band) deformed_min_clearance = std::min(deformed_min_clearance, item[2]);

  EXPECT_GT(deformed_min_clearance, baseline_min_clearance);
}

// (c) an obstacle-free straight corridor: contraction alone keeps the band
// exactly on the line joining its two fixed endpoints --------------------------
TEST(ElasticBands, StraightBandStaysStraight) {
  auto params = real_config();
  auto map = maps::load_map(test::repo_path("maps/grid/open01.yaml"));
  auto& grid = as_grid(*map);
  const Point start_xy{2.0, 8.75};
  const Point goal_xy{8.0, 8.75};  // horizontal segment, obstacle-free within rho_influence
  ElasticBandsPlanner planner(params);
  RobotState state{Pose{start_xy.x, start_xy.y, 0.0}, 0.0, 0.0};
  LocalTask task{Pose{goal_xy.x, goal_xy.y, 0.0}, {start_xy, goal_xy}};

  std::ostringstream os;
  TraceRecorder rec(os);
  planner.compute_command(grid, state, task, params.get_float("control_dt"), &rec);

  std::vector<std::string> band_lines = lines_with_event(split_lines(os.str()), "band_updated");
  ASSERT_FALSE(band_lines.empty());
  std::vector<std::vector<double>> band = parse_band(band_lines.back());
  ASSERT_FALSE(band.empty());
  for (const auto& item : band) EXPECT_NEAR(item[1], start_xy.y, 1e-6);
}

// (d) an unreachable rho_min forces the very first tick's validity check to
// fail: zero command, broken flagged in the trace -------------------------------
TEST(ElasticBands, BandBreakYieldsZeroCommand) {
  auto params = config_with_rho_min(1.2);
  auto map = maps::load_map(test::repo_path("maps/grid/clutter01.yaml"));
  auto& grid = as_grid(*map);
  maps::Scenario sc = maps::load_scenario(test::repo_path("maps/scenarios/clutter01_s3.yaml"));
  ElasticBandsPlanner planner(params);
  RobotState state{Pose{sc.start.x, sc.start.y, sc.start_theta}, 0.0, 0.0};
  LocalTask task{Pose{sc.goal.x, sc.goal.y, sc.goal_theta}, sc.reference_path};

  std::ostringstream os;
  TraceRecorder rec(os);
  core::VelocityCommand cmd = planner.compute_command(grid, state, task, params.get_float("control_dt"), &rec);

  EXPECT_DOUBLE_EQ(cmd.v, 0.0);
  EXPECT_DOUBLE_EQ(cmd.omega, 0.0);
  std::vector<std::string> band_lines = lines_with_event(split_lines(os.str()), "band_updated");
  ASSERT_EQ(band_lines.size(), 1u);
  EXPECT_DOUBLE_EQ(find_data_value(band_lines[0], "broken"), 1.0);
}

// (e) an out-of-range override fails load-time validation -----------------------
TEST(ElasticBands, ParamValidation) {
  std::string bad = test::write_temp(
      "elastic_bands_k_repulsion.yaml",
      "algorithm: elastic_bands\ncategory: local_planning\nparams:\n"
      "  - {name: k_repulsion, type: float, default: -1.0, min: 0.0, max: 10.0, "
      "description: below min}\n");
  EXPECT_THROW(ParamSet::from_yaml(bad), std::runtime_error);
}
